README.md
# Rsyslog mysql housekeeping script

Free as in freedom rsyslog mysql housekeeping script.

The current change log can be found [here](CHANGELOG.md).

The current documentation can be found [here](documentation).

It all started with this question [>>How to automatically delete database contents?<<](https://www.rsyslog.com/article50/).

The script removes entries older than `<configured value>` days from the database table `Syslog.SystemEvents`.

The script comes with an [install](bin/install.sh)- and an [uninstall](bin/uninstall.sh) routine.
All configurable values are configured in `project_path/data/local_configuration.sh`.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Open]

### To Add

* Add support for server when mysql server is not on host `localhost`
* Try to read `/etc/rsyslog.d/mysql.conf` to automatically configure or generate the config file.
* Try to add a arch linux aur package

### To Change

* Hardening the create systemd service by using [this guide](https://www.opensourcerers.org/2022/04/25/optimizing-a-systemd-service-for-security/)
* Refactore systemd timers or service file to pass in the current used configuration?
    * This way we could run a simple "every 15 minutes" cleanup job
    * And a once per week "optimize table" job
    * Think about if it even makes more sense to have two seperate services and timers?
* Add a configurable Limit because of [this post](https://forums.mysql.com/read.php?20,264405,264433#msg-264433)
    * Maybe read the output of the deletion statement like described [here](https://www.pontikis.net/blog/store-mysql-result-to-array-from-bash), or [here](https://www.cloudsavvyit.com/1081/check-a-value-in-a-mysql-database-from-a-linux-bash-script/) and loop until deletion is 0 or less than 1000

## [Unreleased]

### Added

* added `auto_elevate_if_not_called_from_root`

### Changed

## [1.0.1](https://github.com/bazzline/rsyslog_mysql_housekeeping/tree/1.0.1) - released at 20210521

### Changed

* updated timers to use unit `rsyslog-housekeeping.service`
* renamed installed service file from `weekly-rsyslog-housekeeping.service` to `rsyslog-housekeeping.service`
* fixed systemd detection in install and uninstall scripts

## [1.0.0](https://github.com/bazzline/rsyslog_mysql_housekeeping/tree/1.0.0) - released at 20210426

### Added

* Added two different timers plus extended installation script
* Started [documentation](documentation) section
* Added logging
* Added `sleep 10` for every deletion round
* Added configuration values `CONFIGURATION_VERSION`, `EXECUTE_DATABASE_CHECK`, `EXECUTE_DATABASE_OPTIMIZE` and `EXECUTE_DATABASE_ANALYZE`
    * First value is used to handle different configuration versions
    * Other three values are used to fine control database table maintenance

### Changed

* Fixed major issue in `mysqlcheck` section
* Add `NUMBER_OF_ENTRIES_TO_DELETE_PER_RUN` to the configuration used as `LIMIT` because of [this post](https://forums.mysql.com/read.php?20,264405,264433#msg-264433)
    * This will stop having endless running processes
    * This will stop running into socket timeouts
* Added `NUMBER_OF_RUNS=10` to configuration to chunk the total process
* Moved `local_config.sh.dist` from [data](data) to [source](source)
* Changed usage of `mysqlcheck`, it now is working only on `Syslog`.`SystemEvents`, before it was working on the whole database `Syslog`

## [0.9.0](https://github.com/bazzline/rsyslog_mysql_housekeeping/tree/0.9.0) - released at 20210420

### Added

* initial release


install.sh
#!/bin/bash
####
# Installs housekeeping scripts
####
# @since 2021-04-21
# @author stev leibelt <artodeto@bazzline.net>
####

function _install_rsyslog_housekeeping ()
{
    local PATH_OF_THE_CURRENT_SCRIPT_BASH=$(cd $(dirname "${BASH_SOURCE[0]}"); pwd)

    if [[ -f ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/.is_installed ]];
    then
        echo ":: Installation already done."
        echo "   Please run >>${PATH_OF_THE_CURRENT_SCRIPT_BASH}/uninstall.sh<< if needed."

        return 1
    fi

    cp ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/local_config.sh.dist ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/local_config.sh

    echo ":: Please configure following file."
    echo "   >>${PATH_OF_THE_CURRENT_SCRIPT_BASH}/local.config.sh<<"
    echo ""

    if [[ -f /usr/bin/systemd ]];
    then
        local TEMPLATE_PATH_TO_THE_SCRIPT="${PATH_OF_THE_CURRENT_SCRIPT_BASH}/cleanup_and_maintain_syslog_systemevent.sh"

        cat > ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/weekly-rsyslog-housekeeping.service <<DELIM
[Unit]
Description=Weekly rsyslog mysql housekeeping
ConditionACPower=true
After=rsyslog.service

[Service]
Type=oneshot
ExecStart=${TEMPLATE_PATH_TO_THE_SCRIPT}
KillMode=process
TimeoutStopSec=21600
DELIM

        sudo cp ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/weekly-rsyslog-housekeeping.service /etc/systemd/system/weekly-rsyslog-housekeeping.service
        sudo cp ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/weekly-rsyslog-housekeeping.timer /etc/systemd/system/weekly-rsyslog-housekeeping.timer
        sudo systemctl daemon-reload

        echo ":: Please enable the time on your own."
        echo "   systemctl enable weekly-rsyslog-housekeeping.timer"
    fi

    echo "Installed at: $(date)" > ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/.is_installed
}

_install_rsyslog_housekeeping


uninstall.sh
#!/bin/bash
####
# Uninstalls housekeeping scripts
####
# @since 2021-04-21
# @author stev leibelt <artodeto@bazzline.net>
####

function _uninstall_rsyslog_housekeeping ()
{
    local PATH_OF_THE_CURRENT_SCRIPT_BASH=$(cd $(dirname "${BASH_SOURCE[0]}"); pwd)

    if [[ ! -f ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/.is_installed ]];
    then
        echo ":: No Installation found."
        echo "   Please run >>${PATH_OF_THE_CURRENT_SCRIPT_BASH}/install.sh<< if needed."

        return 1
    fi

    if [[ -f /usr/bin/systemd ]];
    then
        systemctl -q is-active weekly-rsyslog-housekeeping.timer

        if [[ $? -eq 0 ]];
        then
            sudo systemctl disable weekly-rsyslog-housekeeping.timer
        fi

        sudo rm /etc/systemd/system/weekly-rsyslog-housekeeping.timer
        sudo rm /etc/systemd/system/weekly-rsyslog-housekeeping.service

        rm ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/weekly-rsyslog-housekeeping.service

        sudo systemctl daemon-reload
    fi

    mv ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/local_config.sh ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/local_config.sh.save

    rm ${PATH_OF_THE_CURRENT_SCRIPT_BASH}/.is_installed
}

_uninstall_rsyslog_housekeeping


cleanup_and_maintain_syslog_systemevent.sh
#!/bin/bash
####
# Deletes entries older than x days
# Optimize table afterwards.
####
# @since 2021-04-21
# @author stev leibelt <artodeto@bazzline.net>
####

####
# @param: string <database user name>
# @param: string <database user password>
# [@param: int <days to keep in the past>] #default is keep the last 90 days
####
function cleanup_and_maintain_syslog_systemevent ()
{
    local DATABASE_NAME="Syslog"
    local DATABASE_PASSWORD="${2:-tb1111}"
    local DATABASE_TABLE="SystemEvents"
    local DATABASE_USER="${1:-syslog}"
    local DAYS_TO_KEEP_IN_THE_PAST="${3:-90}"

    #bo: cleanup
    mysql -u ${DATABASE_USER} -p${DATABASE_PASSWORD} -e "DELETE FROM `${DATABASE_TABLE}` WHERE `${DATABASE_TABLE}`.`DeviceReportedTime` < date_add(current_date, interval - ${DAYS_TO_KEEP_IN_THE_PAST} day)" ${DATABASE_NAME}
    #eo: cleanup

    #bo: maintenance
    #   check table health
    mysqlcheck -u ${DATABASE_USER} -p${DATABASE_PASSWORD} --check --databases ${DATABASE_NAME}
    #   reclaim unused disk space
    mysqlcheck -u ${DATABASE_USER} -p${DATABASE_PASSWORD} --optimize --databases ${DATABASE_NAME}
    #   rebuild and optimize indexes
    mysqlcheck -u ${DATABASE_USER} -p${DATABASE_PASSWORD} --analyze --databases ${DATABASE_NAME}
    #eo: maintenance
}

PATH_OF_THE_CURRENT_SCRIPT_BASH=$(cd $(dirname "${BASH_SOURCE[0]}"); pwd)

PATH_TO_THE_LOCAL_CONFIGURATION_FILE="${PATH_OF_THE_CURRENT_SCRIPT_BASH}/data/local_config.sh"

if [[ -f "${PATH_TO_THE_LOCAL_CONFIGURATION_FILE}" ]];
then
    source "${PATH_TO_THE_LOCAL_CONFIGURATION_FILE}"
fi

cleanup_and_maintain_syslog_systemevent "${DATABASE_USER_NAME}" "${DATABASE_USER_PASSWORD}" ${DAYS_TO_KEEP_IN_THE_PAST}
